server/watchdog/watchdog.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watchdog
+
+import (
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+const (
+	tickMs = 100
+)
+
+var (
+	wdInstance     *watchdog
+	nextActivityId uint64
+)
+
+type activity struct {
+	id              uint64
+	name            string
+	inactiveElapsed int
+}
+
+type watchdog struct {
+	lg *zap.Logger
+
+	ticker              *time.Ticker
+	inactiveTimeoutTick int
+
+	mu         sync.Mutex
+	activities map[uint64]*activity
+
+	stopC   <-chan struct{}
+	cleanup func()
+}
+
+func (wd *watchdog) run() {
+	inactiveTimeout := time.Duration(wd.inactiveTimeoutTick*tickMs) * time.Millisecond
+	wd.lg.Info("Watchdog is running", zap.Duration("inactiveTimeout", inactiveTimeout))
+	for {
+		select {
+		case <-wd.ticker.C:
+			wd.mu.Lock()
+			for _, v := range wd.activities {
+				v.inactiveElapsed++
+				if v.inactiveElapsed > wd.inactiveTimeoutTick/2 {
+					elapsedTime := time.Duration(v.inactiveElapsed*tickMs) * time.Millisecond
+					wd.lg.Warn("Slow activity detected", zap.String("activity", v.name), zap.Duration("duration", elapsedTime))
```

Comment:
nit: please add a gauge metric for slow activities count with activity name as label. I think this will be useful for folks who rely on metrics to actively monitor etcd cluster.